ipv8/dht/discovery.py

```diff
@@ -0,0 +1,185 @@
+import time
+
+from collections import defaultdict
+
+from twisted.internet.defer import fail
+from twisted.internet.task import LoopingCall
+from twisted.python.failure import Failure
+
+from .community import DHTCommunity, Request, PING_INTERVAL, TARGET_NODES, \
+                                            gatherResponses, MAX_NODES_IN_FIND
+from .routing import NODE_STATUS_BAD, Node
+from .payload import StorePeerRequestPayload, StorePeerResponsePayload, \
+                     ConnectPeerRequestPayload, ConnectPeerResponsePayload, \
+                     PingRequestPayload, PingResponsePayload
+
+MSG_STORE_PEER_REQUEST = 13
+MSG_STORE_PEER_RESPONSE = 14
+MSG_CONNECT_PEER_REQUEST = 15
+MSG_CONNECT_PEER_RESPONSE = 16
+
+
+class DHTDiscoveryCommunity(DHTCommunity):
+    """
+    Community for discovering peers that are behind NAT.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DHTDiscoveryCommunity, self).__init__(*args, **kwargs)
+
+        self.store = defaultdict(list)
+        self.store_for_me = defaultdict(list)
+
+        self.decode_map.update({
+            chr(MSG_STORE_PEER_REQUEST): self.on_store_peer_request,
+            chr(MSG_STORE_PEER_RESPONSE): self.on_store_peer_response,
+            chr(MSG_CONNECT_PEER_REQUEST): self.on_connect_peer_request,
+            chr(MSG_CONNECT_PEER_RESPONSE): self.on_connect_peer_response,
+        })
+
+        self.register_task('store_peer', LoopingCall(self.store_peer)).start(30, now=False)
+
+    def on_ping_request(self, source_address, data):
```

Comment:
These two methods differ only in a couple of words. Maybe, generalize?

Reply:
I'll take a look.